proposals/UIWebView-strategy.md

```diff
@@ -0,0 +1,86 @@
+UIWebView Strategy for Cordova iOS
+==================================
+
+As of August 2019, Apple is now showing a deprecation warning when uploading apps to the App Store that include UIWebView-related code. UIWebView has been unofficially deprecated in favour of the new WKWebView for a while, but Apple has now made it officially deprecated. As a result, all Cordova apps built for iOS will receive this deprecation warning on upload. Currently it is just a warning, but Apple has not provided a timeline for when it will start rejecting apps that use UIWebView APIs.
+
+We've had [a plan](https://cordova.apache.org/news/2018/08/01/future-cordova-ios-webview.html) for several years to make it easier to migrate to WKWebView and eventually deprecate UIWebView use by Cordova, but almost no progress has been made on this objective since the plan was outlined over a year ago.
+
+Cordova users are clamoring for a fix to the problem and there are pull requests open to remove UIWebView entirely, but if we proceed without a concrete plan then we are guaranteed to cause a lot of pain and suffering for Cordova users.
+
+The goal of this document is to outline several potential migration plans, and facilitate discussion on the pros and cons of each.
+
+
+High-Level Goals
+----------------
+
+* Make WKWebView the default web view for Cordova iOS applications
+* Make it possible to build Cordova iOS apps without including any UIWebView APIs
+* Maintain compatibility as much as possible with existing plugins
+* Provide a means for Cordova apps to transition from UIWebView to WKWebView
+
+
+Challenges
+----------
+
+* UIWebView-related classes are potentially part of the public API exposed to plugins
+* Moving all apps over to WKWebView will result in data being lost (localStorage, indexedDB, etc.)
+* WKWebView does not support all the same features as UIWebView and has stricter security requirements, and **many existing Cordova iOS apps will not work if run in WKWebView** without changes
+* The current WKWebView plugin needs updates to use Scheme Handlers to avoid issues that require a local web server.
+* The InAppBrowser plugin also uses UIWebView. Moving to WKWebView would potentially lose some functionality.
```

Comment:
InAppBrowser has supported WKWebview for quite some time via `usewkwebview`.

Reply:
IAB with WKWebView basically works fine now except this issue https://github.com/apache/cordova-plugin-inappbrowser/issues/492

IAB also needs to hide UIWebView behind a build flag, like some other plugins.

Reply:
What about LocalStorage from HTML5? That stuff is used all over the place at least by many apps I've made with Cordova. We can always make a backend system to send XMLHttpRequests to store data, but LocalStorage is just so much faster for small apps.

If I want to help make LocalStorage a possibility, would cordova-ios be the project to try to contribute to, or try to figure out solutions here?

Reply:
I am not sure if LocalStorage should be part of this discussion. LocalStorage should work in WKWebView just fine. The only problem is that switching from UIWebView to WKWebView makes your apps data inaccessible with the new webview. There are some solutions out there to migrate. But this is not a new problem since WKWebView has been around some time. 

I don't know if this migration should be part of cordova-ios.

Reply:
They can also use SafariViewController for inAppBrowser

Reply:
Just to play devil's advocate here, on the topic of local storage...

There are some plugins out there that migrate local storage such as [this](https://www.npmjs.com/package/cordova-plugin-migrate-localstorage) one, but most of them will only migrate if the `WkWebView` storage containers are empty (otherwise they run into a question of "should they overwrite?").

If one would update cordova-ios and thus migrate to the wkwebview without the use of this plugin, they may not realise that they lost data, and by that time it's too late to use a migration plugin. To save headaches for developers, maybe we should incorporate of these plugins directly into cordova-ios.

But on the other side, we could just simply tell users to consider using one of those plugins so that we don't have to maintain an one-time use code for an entire major version lifetime.

Personally I think I'm still on the "not cordova's problem" side...